src/appleseed.studio/mainwindow/rendering/lightpathswidget.cpp

```diff
@@ -160,149 +181,471 @@ void LightPathsWidget::slot_synchronize_camera()
+        }
     }
 
-    void glColor(const Color3f& c)
+    void create_shader_program(
+        GLuint&                 program,
+        const std::string&      vert_source,
+        const std::string&      frag_source)
     {
-        glColor3f(c[0], c[1], c[2]);
+
+        GLuint vert = glCreateShader(GL_VERTEX_SHADER);
+        GLuint frag = glCreateShader(GL_FRAGMENT_SHADER);
+
+        auto gl_vert_source = static_cast<const GLchar*>(vert_source.c_str());
+        auto gl_frag_source = static_cast<const GLchar*>(frag_source.c_str());
+
+        compile_shader(vert, 1, &gl_vert_source, NULL);
+        compile_shader(frag, 1, &gl_frag_source, NULL);
+
+        program = glCreateProgram();
+        link_shader_program(program, vert, frag);
+
+        glDeleteShader(vert);
+        glDeleteShader(frag);
     }
 
+
     struct OpenGLRasterizer
-      : public ObjectRasterizer
+        : public ObjectRasterizer
     {
-        void begin_object() override
-        {
-            glBegin(GL_TRIANGLES);
-        }
+        std::vector<float> buffer;
+        size_t             prim_count;
 
-        void end_object() override
+        void begin_object(const size_t triangle_count_hint) override
         {
-            glEnd();
+            buffer.clear();
+            buffer.reserve(triangle_count_hint * 18);
+            prim_count = 0;
         }
 
+        void end_object() override {}
+
         void rasterize(const Triangle& triangle) override
         {
-            glNormal3d(triangle.m_n0[0], triangle.m_n0[1], triangle.m_n0[2]);
-            glVertex3d(triangle.m_v0[0], triangle.m_v0[1], triangle.m_v0[2]);
-
-            glNormal3d(triangle.m_n1[0], triangle.m_n1[1], triangle.m_n1[2]);
-            glVertex3d(triangle.m_v1[0], triangle.m_v1[1], triangle.m_v1[2]);
-
-            glNormal3d(triangle.m_n2[0], triangle.m_n2[1], triangle.m_n2[2]);
-            glVertex3d(triangle.m_v2[0], triangle.m_v2[1], triangle.m_v2[2]);
+            const float temp_store[18] = {
+                static_cast<float>(triangle.m_v0[0]), static_cast<float>(triangle.m_v0[1]), static_cast<float>(triangle.m_v0[2]),
+                static_cast<float>(triangle.m_n0[0]), static_cast<float>(triangle.m_n0[1]), static_cast<float>(triangle.m_n0[2]),
+
+                static_cast<float>(triangle.m_v1[0]), static_cast<float>(triangle.m_v1[1]), static_cast<float>(triangle.m_v1[2]),
+                static_cast<float>(triangle.m_n1[0]), static_cast<float>(triangle.m_n1[1]), static_cast<float>(triangle.m_n1[2]),
+
+                static_cast<float>(triangle.m_v2[0]), static_cast<float>(triangle.m_v2[1]), static_cast<float>(triangle.m_v2[2]),
+                static_cast<float>(triangle.m_n2[0]), static_cast<float>(triangle.m_n2[1]), static_cast<float>(triangle.m_n2[2]),
+            };
+            buffer.reserve(buffer.size() + 18);
+            buffer.insert(buffer.end(), temp_store, temp_store + 18);
```

Comment:
That is what happens :) The flow is that 

`initializeGL` (once at beginning) -> `load_scene_data` -> `load_assembly_data` -> `load_object_data` -> if this object mesh hasn't yet been uploaded, create an OpenGLRasterizer and rasterize it to the internal buffer, then upload that to the GPU. The buffer reserves enough capacity for the `triangle_count_hint` in `OpenGLRasterizer::begin_object`, however I am being conservative here and still adding the buffer reserve/insert in case in the future there is another kind of object which cannot (or can't efficiently) give an exact triangle count at the beginning.

`paintGL` (each frame) -> `render_geometry` -> iterate through all (already uploaded) vertex buffers and issue a draw command on all of the uploaded instances